src/dataflow/src/render/upsert.rs

```diff
@@ -398,18 +401,67 @@ where
                                     // We store errors as well as non-None values, so that they can be
                                     // retracted if new rows show up for the same key.
                                     let new_value = decoded_value.transpose();
+
+                                    use dataflow_types::plan::Permutation;
+                                    let new_value = new_value.map(|actual_value| {
+                                        actual_value.map(|non_err_val| {
+                                            let total_len = non_err_val.iter().count();
+                                            let (permutation, thinning) =
+                                                Permutation::construct_from_columns(
+                                                    &(0..decoded_key.iter().count())
+                                                        .into_iter()
+                                                        .collect::<Vec<usize>>(),
+                                                    total_len,
+                                                );
```

Comment:
I think we should be able to hoist this logic outside of various loops. If for some reason we don't statically know the number of columns in each key and value, we should figure out how to fix that.